### ISSUE TITLE
Kubelet often fails on AWS spot instances

### ISSUE THREAD

**Issue by vainu-arto** [kubernetes/kubernetes#57382]
/kind bug

In a kops created cluster running Kubernetes 1.8.4 nodes sometimes start but fail to connect to the masters. I have tracked the issue down to this difference:

Failed:
I1216 07:28:57.369519 1900 aws.go:847] Building AWS cloudprovider
I1216 07:28:57.369555 1900 aws.go:810] Zone not specified in configuration file; querying AWS metadata service
E1216 07:28:57.586712 1900 tags.go:94] Tag "KubernetesCluster" nor "kubernetes.io/cluster/..." not found; Kubernetes may behave unexpectedly.
W1216 07:28:57.586744 1900 tags.go:78] AWS cloud - no clusterID filtering applied for shared resources; do not run multiple clusters in this AZ.
I1216 07:28:57.586807 1900 server.go:301] Successfully initialized cloud provider: "aws" from the config file: ""

Successful:
I1218 08:32:38.492573 28476 aws.go:847] Building AWS cloudprovider
I1218 08:32:38.492596 28476 aws.go:810] Zone not specified in configuration file; querying AWS metadata service
I1218 08:32:38.669713 28476 tags.go:76] AWS cloud filtering on ClusterID: kube.test.int
I1218 08:32:38.669788 28476 server.go:301] Successfully initialized cloud provider: "aws" from the config file: ""

So in the failing case the AWS metadata either doesn't respond or the response is missing critical information (I'm not sure how to find out which). This query isn't retried during kubelets general "retry connecting to the api" loop, so the kubelet will keep running but it will never retrieve the information it needs to install and configure calico and join the network.

**Comment by vainu-arto** [comment]
This happens mainly on spot instances. Perhaps at least those don't actually have their tags set at startup, but get them slightly later? In any case it seems the logic of only fetching the metadata once in kubelet and never retrying is probably incorrect.

**Comment by FrederikNS** [comment]
We're running into this issue as well, using Kops 1.8.0 and Kuberntes 1.8.4.
If new nodes fail to add themselves to the cluster, but a restart of the node fixes the issue.

**Comment by amalucelli** [comment]
I've the same issue here.

```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.5", GitCommit:"cce11c6a185279d037023e02ac5249e14daa22bf", GitTreeState:"clean", BuildDate:"2017-12-07T18:09:00Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

Restarting the `kubelet` service on the instance resolve the problem, but it's unpractical doing that to every new spot instance added (at least for me).

I observed that it doesn't happen with every new spot instance added. Maybe those instances that doesn't face this issue can have their tags allocated in time to `kubelet` service startup as mencioned by @vainu-arto.

Is there any workaround for this situation? Other than restarting the `kubelet` service?

**Comment by kruczjak** [comment]
Hi, also same issue, but on Kubernetes 1.8.5. Only restarting `kubelet` worked for me.

**Comment by sdlarsen** [comment]
Same here. On Kubernetes 1.8.4, 1.8.5 and 1.8.6.

**Comment by vainu-arto** [comment]
Another duplicate kops ticket https://github.com/kubernetes/kops/issues/3605 contains a workaround based on a script that polls the api until the tags appear. Kubelet should probably do this itself.

**Comment by dims** [comment]
cc @kubernetes/sig-aws-misc

**Comment by chrislovecnm** [comment]
Personally I think this is an installer problem. The node technically is not ready for k8s. Until the tags exist k8s should not be installed or started.

**Comment by vainu-arto** [comment]
I'd say there are problems on both sides. The installer should not start kubelet before the system is ready, and kubelet should not start and claim success in a situation where it cannot function.

**Comment by jberkus** [comment]
@justinsb if this issue is going to be fixed in 1.10, can you please add a priority/ label to it?  Thanks.

**Comment by jberkus** [comment]
This issue will be automatically removed from 1.10 tommorrow unless you add a priority/ label to it, thanks!

**Comment by tpepper** [comment]
@justinsb please also add approved-for-milestone here and in #60125 if you're wanting these to not get bumped from 1.10 now that we're at code freeze

**Comment by jberkus** [comment]
oops, missing priority was my fault, moving back into milestone:

/priority critical-urgent
/remove-priority important-soon

**Comment by vainu-arto** [comment]
@jberkus The fix to this was merged into master, and also exists in the release-1.10 branch. Is there an action that needs to be taken or should this issue be closed as done instead?

**Comment by jberkus** [comment]
@vainu-arto that's really up to you.  if the issue is now fixed, then close it.

**Comment by vainu-arto** [comment]
This was fixed for 1.10 by PR #60125. Cherrypicks to 1.8 and 1.9 are still pending (#60480 and #60481).
